javaagent-tooling/src/testPatchBytecodeVersion/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/PatchBytecodeVersionTest.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling.instrumentation.indy;
+
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import net.bytebuddy.agent.ByteBuddyAgent;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.agent.builder.ResettableClassFileTransformer;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.dynamic.ClassFileLocator;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.utility.JavaModule;
+import org.apache.commons.lang3.StringUtils;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.objectweb.asm.Opcodes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class PatchBytecodeVersionTest {
+
+  private static final String BYTEBUDDY_DUMP = "net.bytebuddy.dump";
+  private static ResettableClassFileTransformer transformer;
+
+  private static final Logger logger = LoggerFactory.getLogger(PatchBytecodeVersionTest.class);
+
+  private static Path tempDir;
+
+  @BeforeAll
+  static void setUp(@TempDir Path temp) {
+
+    assertThat(temp).isEmptyDirectory();
+    tempDir = temp;
+    System.setProperty(BYTEBUDDY_DUMP, temp.toString());
+
+    AgentBuilder builder =
+        new AgentBuilder.Default()
+            .disableClassFormatChanges()
+            .with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
+            .with(
+                new AgentBuilder.Listener.Adapter() {
+                  @Override
+                  public void onError(
+                      String typeName,
+                      ClassLoader classLoader,
+                      JavaModule module,
+                      boolean loaded,
+                      Throwable throwable) {
+                    logger.error("Transformation error", throwable);
+                  }
+                })
+            // commons lang 3
```

Comment:
Instead of looking for java libs that were compiled against a particular class file version, WDYT about generating these classes in the runtime? We already have a somewhat similar test that verifies that an instrumentation works on Java6: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/dd2bccdd3ea4c1a48aecea7fe60c4646b21feaa0/instrumentation/opentelemetry-instrumentation-annotations-1.16/javaagent/src/test/java/io/opentelemetry/test/annotation/WithSpanInstrumentationTest.java#L510-L531

If you use the same approach you can make a parameterized test that tests pretty much all the older class file versions before 8.

Reply:
Thanks a lot for the link, I will attempt to implement that in this PR as it's probably the best compromize.

So far the alternatives I was thinking of were:
- using known existing old libraries (what this PR currently does, not really ideal)
- using pre-compiled binaries. Not ideal for lack of transparency and harder to maintain over time
- compiling with an old JDK, Gradle does not seem to support that, and this would further complicate development if an old JDK is required.

Reply:
I've implemented this in 4125f2e58b6f82b338141a5399dc5233d6748a40 by generating old bytecode with bytebuddy.